File: 967-minimum-falling-path-sum/minimum-falling-path-sum.cpp
class Solution {
public:
    int memo(vector<vector<int>>& matrix, int i, int j, vector<vector<int>>& dp){
        int n = matrix.size();
        int m = matrix[0].size();

        if(i < 0 || j < 0 || j >= m) return INT_MAX;

        if(i == 0) return matrix[i][j];

        if(dp[i][j] != INT_MAX) return dp[i][j];

        long long d = matrix[i][j] * 1LL + memo(matrix, i - 1, j, dp);
        long long rd = matrix[i][j] * 1LL + memo(matrix, i - 1, j - 1, dp);
        long long ld = matrix[i][j] * 1LL + memo(matrix, i - 1, j + 1, dp);

        return dp[i][j] = (int)min({d, rd, ld});
        
    }

    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();

        vector<vector<int>> dp(n, vector<int>(m, INT_MAX));

        int ans = INT_MAX;

        for(int j = 0; j < m; j++){
            ans = min(ans, memo(matrix, n - 1, j, dp)); 

        }

        return ans;
    }
};

File: 967-minimum-falling-path-sum/minimum-falling-path-sum.java
class Solution {
    int dp[][];
    private int solve(int[][] matrix, int i, int j) {
        int n = matrix.length;
        
        if (i >= n) return 0;

        if(dp[i][j] != Integer.MIN_VALUE) return dp[i][j];

        int op1 = solve(matrix, i + 1, j);
        int op2 = Integer.MAX_VALUE;
        int op3 = Integer.MAX_VALUE;

        if (j - 1 >= 0) op2 = solve(matrix, i + 1, j - 1);
        if (j + 1 <= n - 1) op3 = solve(matrix, i + 1, j + 1);

        return dp[i][j] = matrix[i][j] + Math.min(Math.min(op1, op2), op3);
    }

    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;

        dp = new int[n][m];
        
        for (int[] row : dp) {
            Arrays.fill(row, Integer.MIN_VALUE);
        }

        int minSum = Integer.MAX_VALUE;

        for (int j = 0; j < n; j++) {
            minSum = Math.min(minSum, solve(matrix, 0, j));
        }

        return minSum;
    }
}